### ISSUE TITLE
After publishing blazor server service model in the client gives argumentoutofrangeexception

### ISSUE THREAD

**Issue by linmic30** [max-ieremenko/ServiceModel.Grpc#204]
I have a wasm blazor app that communicates using grpc and the servicemodel library to create the interfaces for GRPC. when i publish the server side of this app to deploy i get the following error in the browser:

``` 
Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Arg_ArgumentOutOfRangeException Arg_ParamName_Name, Method`2 has no .ctor (MethodType,String,String,Marshaller`1,Marshaller`1).
System.ArgumentOutOfRangeException: Arg_ArgumentOutOfRangeException Arg_ParamName_Name, Method`2 has no .ctor (MethodType,String,String,Marshaller`1,Marshaller`1).
   at ServiceModel.Grpc.Internal.ReflectionTools.Constructor(Type type, Type[] parameters)
   at ServiceModel.Grpc.Internal.Emit.EmitContractBuilder.BuildMethod(OperationDescription operation, TypeBuilder typeBuilder, ILGenerator ctor)
   at ServiceModel.Grpc.Internal.Emit.EmitContractBuilder.Build(ModuleBuilder moduleBuilder, String className)
   at ServiceModel.Grpc.Internal.Emit.EmitGenerator.GenerateContract(Type serviceType)
   at ServiceModel.Grpc.Internal.Emit.EmitGenerator.GenerateClientBuilder[IWeatherForecastService]()
   at ServiceModel.Grpc.Client.ClientFactory.RegisterClient[IWeatherForecastService](IClientBuilder`1 userBuilder, Action`1 configure)
   at ServiceModel.Grpc.Client.ClientFactory.CreateClient[IWeatherForecastService](CallInvoker callInvoker)
   at ServiceModel.Grpc.Client.ClientFactory.CreateClient[IWeatherForecastService](ChannelBase channel)
   at BlazorApp.Client.Program.<>c.<ConfigureServices>b__1_2(IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite , RuntimeResolverContext )
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSiteMain(ServiceCallSite , RuntimeResolverContext )
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite , RuntimeResolverContext , ServiceProviderEngineScope , RuntimeResolverLock )
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite , RuntimeResolverContext )
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSite(ServiceCallSite callSite, RuntimeResolverContext argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite , ServiceProviderEngineScope )
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeServiceProviderEngine.<>c__DisplayClass4_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier , ServiceProviderEngineScope )
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type )
   at Microsoft.AspNetCore.Components.ComponentFactory.<>c__DisplayClass9_0.<CreatePropertyInjector>g__Initialize|1(IServiceProvider serviceProvider, IComponent component)
   at Microsoft.AspNetCore.Components.ComponentFactory.InstantiateComponent(IServiceProvider , Type , IComponentRenderMode , Nullable`1 )
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.InstantiateChildComponentOnFrame(RenderTreeFrame[] , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InitializeNewComponentFrame(DiffContext& , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InitializeNewSubtree(DiffContext& , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.InsertNewFrame(DiffContext& , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& , Int32 , Int32 , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& , Int32 , Int32 , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& , Int32 , Int32 , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& , Int32 , Int32 , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& , Int32 , Int32 , Int32 , Int32 )
   at Microsoft.AspNetCore.Components.RenderTree.RenderTreeDiffBuilder.ComputeDiff(Renderer , RenderBatchBuilder , Int32 , ArrayRange`1 , ArrayRange`1 )
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder , RenderFragment , Exception& )
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.RenderInExistingBatch(RenderQueueEntry )
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.ProcessRenderQueue()
```

This happens when the first GRPC call needs to be made and the creation of the interface classes is started. I have this with my own app and the demo app in the repo you provide on GITHUB. I tried this with and without selfcontaind features in Visual Studio.

**To Reproduce**
right click on the server project and publish and publish to folder. Start the app as administrator and go to the website an initiate a GRPC call.

**Expected behavior**
The behaviour i expect is that the GRPC interface is setup and the call is made.

**Environment**
 - OS: windows 10/11 windows server 2022
 - .net 8

**Additional context**
It looks like no controctor can be found when it looks for it in the ServiceModel.Grpc.Internal.ReflectionTools.Constructor function. It is only unclear for me where the constructor should be. Also i added the following to the client project to prevent trimming of the different dll's in the client project, but this also did not help:
```
  <ItemGroup>
    <TrimmerRootAssembly Include="System.ServiceModel.Primitives" />
    <TrimmerRootAssembly Include="Grpc.Net.Client" />
    <TrimmerRootAssembly Include="Grpc.Net.Client.Web" />
    <TrimmerRootAssembly Include="ServiceModel.Grpc" />
    <TrimmerRootAssembly Include="ServiceModel.Grpc.ProtoBufMarshaller" />
  </ItemGroup>
```

**Comment by max-ieremenko** [comment]
Thank you for the report.

It is a trimming issue, after publishing, in the browser type `Grpc.Core.Method<,>` does not have the ctor.

As a temporary solution, you can hint not to trim Grpc.Core.Method, for example, mark [ConfigureServices method](https://github.com/max-ieremenko/ServiceModel.Grpc/blob/954391c3d9b36dfcebacb4968111ae2b3152c409/Examples/BlazorApp/Client/Program.cs#L24) in the Program.cs with DynamicDependency

```c#
[DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(Method<,>))]
private static void ConfigureServices(IServiceCollection services)
```

or place the attribute elsewhere.

The fix will come soon with the next release.

**Comment by linmic30** [comment]
Thank you for the workaround, it seems to be working. 

When do you expect the new release? Then this workaround is maybe not necessary anymore.

**Comment by max-ieremenko** [comment]
The plan is to release this fix at the end of this week.

**Comment by max-ieremenko** [comment]
The issue is fixed in [1.8.3](https://github.com/max-ieremenko/ServiceModel.Grpc/releases/tag/1.8.3) version.
